### PR TITLE
Adds a max length config option for AI law upload boards.

### PR DESCRIPTION
**Pull request by zxaber**
## About The Pull Request
Adds character limit config option for AI upload boards.

## Why It's Good For The Game
image

Laws should be encouraged to be clear and concise. At the very least, having to decipher a paragraph in the middle of doing anything is terrible, and there's not always an admin on or willing to take the appropriate action when someone is making paragraph laws (which would be immediately and unconditionally granting the AI malf status).

Asimov board was missing it's max name setting, so I fixed that too to be inline with the other similar boards (like one-human).

## Changelog
:cl: Zxaber
tweak: There is now a max character config option for AI Freeform, Core Freeform, and Asimov upload boards.
/:cl:

### CODE REVIEW

**Comment by wesoda25** [comment]
God I love laws like this

**Comment by QualityVan** [comment]
Should be 140 instead.

**Comment by zxaber** [comment]
Why 140, exactly? I'm open to discussion about it, but I'm curious why that number specifically.

**Review by ExcessiveUseOfCobblestone — CHANGES_REQUESTED** [review]
define the length or use preexisting one

https://github.com/tgstation/tgstation/blob/a9ca93b60dbf7b9f0fc5cbe94c406a2126312fee/code/__DEFINES/say.dm#L76-L80

**Review by ShizCalev — CHANGES_REQUESTED** [review]
make it a config option. i know a couple servers have lawsets that easily pass 250 chars per law. 

heck, one of the standard AI law boards even reach that limit - https://github.com/tgstation/tgstation/blob/c4213f13ab6ab368d00db9f4383bdf556c1bc4ea/code/game/objects/items/AI_modules.dm#L218-L220

**Comment by ATH1909** [comment]
I've already hit the cap on law length before without trying to with the current length; I'd imagine that the length limit would become much more annoying if this change went through.

Reducing the maximum length of AI laws won't solve this problem; people will just split their mega-laws into smaller (but just as long when their lengths are all added together) laws, resulting in arguably worse law-spam.

Since that's kind of a weird sentence, I'll show what I mean:
Before this change:
1. What the fuck did you just fucking say about me, you little bitch? I'll have you know I graduated top of my class in the Navy Seals, and I've been involved in numerous secret raids on Al-Quaeda, and I have over 300 confirmed kills. I am trained in gorilla warfare and I'm the top sniper in the entire US armed forces. You are nothing to me but just another target. I will wipe you the fuck out with precision the likes of which has never been seen before on this Earth, mark my fucking words. You think you can get away with saying that shit to me over the Internet? Think again, fucker. As we speak I am contacting my secret network of spies across the USA and your IP is being traced right now so you better prepare for the storm, maggot. The storm that wipes out the pathetic little thing you call your life. You're fucking dead, kid. I can be anywhere, anytime, and I can kill you in over seven hundred ways, and that's just with my bare hands. Not only am I extensively trained in unarmed combat, but I have access to the entire arsenal of the United States Marine Corps and I will use it to its full extent to wipe your miserable ass off the face of the continent, you little shit. If only you could have known what unholy retribution your little "clever" comment was about to bring down upon you, maybe you would have held your fucking tongue. But you couldn't, you didn't, and now you're paying the price, you goddamn idiot. I will shit fury all over you and you will drown in it. You're fucking dead, kiddo.

After this change:
1. What the fuck did you just fucking say about me, you little bitch? I'll have you know I graduated top of my class in the Navy Seals, and I've been involved in numerous secret raids on Al-Quaeda, and I have over 300 confirmed kills.
2. I am trained in gorilla warfare and I'm the top sniper in the entire US armed forces. You are nothing to me but just another target. I will wipe you the fuck out with precision the likes of which has never been seen before on this Earth, mark my fucking words.
3. You think you can get away with saying that shit to me over the Internet? Think again, fucker. As we speak I am contacting my secret network of spies across the USA and your IP is being traced right now so you better prepare for the storm, maggot.
4. The storm that wipes out the pathetic little thing you call your life. You're fucking dead, kid. I can be anywhere, anytime, and I can kill you in over seven hundred ways, and that's just with my bare hands. 
5. Not only am I extensively trained in unarmed combat, but I have access to the entire arsenal of the United States Marine Corps and I will use it to its full extent to wipe your miserable ass off the face of the continent, you little shit.
6. If only you could have known what unholy retribution your little "clever" comment was about to bring down upon you, maybe you would have held your fucking tongue.
7. But you couldn't, you didn't, and now you're paying the price, you goddamn idiot. I will shit fury all over you and you will drown in it. You're fucking dead, kiddo.

**Comment by zxaber** [comment]
ATH1909, the current limit is (I believe) just over a thousand characters. If you're making thousand-character laws, you need to stop.

Fifteen one-line laws are more readable and easier to deal with than one fifteen-line law.

If you're hitting the limit on boards like the one-human, those have limit that was already set, and this PR doesn't change that (except that it adds the same limit to the Asimov board).

**Comment by optimumtact** [comment]
this is a lame change

**Comment by zxaber** [comment]
Getting a law in the form of an essay is pretty lame too. It's not that hard to condense ideas down, unless the uploader is intentionally trying to cause a headache.

**Comment by necromanceranne** [comment]
That 5th law deserves granting the AI malf. God what cancer.

**Comment by HomocidlePancakes** [comment]
>Getting a law in the form of an essay is pretty lame too. It's not that hard to condense ideas down, >unless the uploader is intentionally trying to cause a headache.
>>the uploader is intentionally trying to cause a headache.
>>the uploader is intentionally trying to cause a headache.

HALP I DONT LIKE WHEN PEPPL GRIFF 

The intention of your PR is to limit amount of grief people do with laws, but I don't see how a char limit will in any way restrict that. They can easily break it down in to multiple laws (and it will not always make it easier to read). You can still force AI to do dumb shit in one law to the same effect as shitting out an entire essay. Heck you can even crash silicons by spamming upload over and over. This PR liturally accomplishes nothing.

**Comment by AlexTheSysop42** [comment]
When laws are updated it redisplays all your laws iirc so more small laws just means extra chatspam for the ai

**Comment by ExcessiveUseOfCobblestone** [comment]
Why not just make it the broadcast limit? That should be fine for a vast majority of laws and is about the same level of obnoxious when misused.

**Comment by zxaber** [comment]
Using a define from a mostly-unrelated feature would seem a bit hacky. If the broadcast limit ever changed for whatever reason, it would change here too and might be unnoticed. I could change it to match the current broadcast limit, but this is proving to be a rather unpopular PR so ¯\_(ツ)_/¯

**Comment by ExcessiveUseOfCobblestone** [comment]
I'm more concerned about https://github.com/tgstation/tgstation/blob/a9ca93b60dbf7b9f0fc5cbe94c406a2126312fee/code/__DEFINES/say.dm#L76

If it's still true.

**Comment by MrPerson** [comment]
You would be surprised how much easier it is to read and parse that 10 law copypasta vs one paragraph with the same amount of text. There's a point where any law you're typing is too long and it's much shorter than 1000 characters.

**Comment by Anonmare** [comment]
If your law needs a flowchart longer than three steps to resolve itself then it's too fucking long and the Ai gets a free pass to ignore it

**Comment by AlexTheSysop42** [comment]
The best part I don't think the law 5 in the screenshot actually changes things

**Comment by lordsaladin** [comment]
@zxaber kinda ruins copypasta laws, instead long ass laws should make a ai go malf.

**Comment by YakumoChen** [comment]
long-ass laws should open up admins to giving AI malf, more like. We shouldn't have a setting that is automatically granting the AI antag status.

**Comment by necromanceranne** [comment]
Yeah uh let's not just allow any old idiot the ability to give the AI malf status. That is gonna get abused for self-antagging and validhunting like fucking crazy.

**Comment by ExcessiveUseOfCobblestone** [comment]
operative?

**Comment by zxaber** [comment]
Right, so;
I currently have two maintainers asking for conflicting changes. Do I continue further on the path of it being a define? Do I make it a config option? I am told I cannot do both.

**Comment by Jordie0608** [comment]
Config.

**Comment by SouDescolado** [comment]
If there is a real need to reduce spam via AI law for the game to be fun, ask for "clarification" when uploading if it has more than 100 characters or something, either like "I'm the only human and that's it" or "This is just a meme plz ignore"

**Comment by ExcessiveUseOfCobblestone** [comment]
> Config.

**Comment by optimumtact** [comment]
To end any debate on the matter, make it a configuration option.

**Comment by zxaber** [comment]
Alright, it's now a config option, defaulting to the existing thesis-length 1024 character limit. Thus, no change occurs directly from this PR alone. If merged, I suppose it becomes something to take to the policy subforum.

File: code/game/objects/items/AI_modules.dm [hunk 1]
```diff
@@ -365,7 +365,7 @@ AI MODULES
 	var/subject = "human being"
 
 /obj/item/aiModule/core/full/asimov/attack_self(var/mob/user as mob)
-	var/targName = stripped_input(user, "Please enter a new subject that asimov is concerned with.", "Asimov to whom?", subject)
+	var/targName = stripped_input(user, "Please enter a new subject that asimov is concerned with.", "Asimov to whom?", subject, MAX_NAME_LEN)
```
Comment (ExcessiveUseOfCobblestone) [inline code/game/objects/items/AI_modules.dm]:
```suggestion
	var/targName = stripped_input(user, "Please enter a new subject that asimov is concerned with.", "Asimov to whom?", subject, CONFIG_GET(number/max_law_len))
```

Comment (zxaber) [reply code/game/objects/items/AI_modules.dm]: Should that one be set to the law length over name length? Other similar boards like the One Human use max name.

Comment (ExcessiveUseOfCobblestone) [reply code/game/objects/items/AI_modules.dm]: reading... in 2019?

**Pull request merged**